### PR TITLE
Obscure client ID in Kinesis

### PR DESCRIPTION
**Pull request by lailasharshar**
**JIRA Tickets:**

[AB2D-2772](https://jira.cms.gov/browse/AB2D-2772) - Make 

### What Does This PR Do?

Before we send an event to Kinesis, obscure the user id, which is the Okta Client ID, by generated an MD5 value of that ID.

### Impacted External Components

BFD Insights

### Security Implications

Reduces visibility of the Client ID in BFD Insights in response to the SIA.

- [ ] This PR adds new software dependencies
- [ ] This PR modifies or invalidates our security controls
- [ ] This PR stores or transmits data that was not stored or transmitted before

- [ ] This PR requires additional review of its security implications for other reasons

### Submitter Checklist

- [ ] This PR includes any required documentation changes, (`README`, changelog / release notes, website).
- [ ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments.
- [ ] Code checked for PHI/PII exposure

### CODE REVIEW

**Review by sb-benohe — CHANGES_REQUESTED** [review]
Request that hash is change to at least SHA256, Recommend SHA3_256

**Review by wnyffenegger — CHANGES_REQUESTED** [review]

File: eventlogger/src/main/java/gov/cms/ab2d/eventlogger/LoggableEvent.java [hunk 1]
```diff
@@ -8,7 +8,7 @@
  * Interface describing a loggable event
  */
 @Data
-public abstract class LoggableEvent {
+public abstract class LoggableEvent implements Cloneable {
```
[the remark this answers is not in the document]
Comment (lailasharshar) [reply eventlogger/src/main/java/gov/cms/ab2d/eventlogger/LoggableEvent.java:11]: Too many items for copy constructors for it and it's super class.

File: eventlogger/src/main/java/gov/cms/ab2d/eventlogger/eventloggers/kinesis/KinesisEventProcessor.java [hunk 2]
```diff
@@ -83,7 +84,11 @@ public static String getJsonString(LoggableEvent event) {
     public Void call() {
         String json = null;
         try {
-            json = getJsonString(event) + "\n";
+            LoggableEvent newLogEvent = event.clone();
+            if (event.getUser() != null && !event.getUser().isEmpty()) {
+                newLogEvent.setUser(DigestUtils.md5Hex(event.getUser()).toUpperCase());
```
[the remark this answers is not in the document]
Comment (lailasharshar) [reply eventlogger/src/main/java/gov/cms/ab2d/eventlogger/eventloggers/kinesis/KinesisEventProcessor.java:89]: It doesn't have to be complicated. It's going to go to a BFD Insights S3 bucket which is already secure and we want to be able to easily figure out what it is.

File: eventlogger/src/main/java/gov/cms/ab2d/eventlogger/LoggableEvent.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (lailasharshar) [reply eventlogger/src/main/java/gov/cms/ab2d/eventlogger/LoggableEvent.java:11]: Also, I'm not cloning without knowing which type of event it is.

File: eventlogger/src/main/java/gov/cms/ab2d/eventlogger/eventloggers/kinesis/KinesisEventProcessor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lailasharshar) [reply eventlogger/src/main/java/gov/cms/ab2d/eventlogger/eventloggers/kinesis/KinesisEventProcessor.java:89]: I actually want to create a dictionary table (with the user's org name, not client id), no one inside of BFD insights is doing a brute force attack and we're not going to have collisions with this few number of client IDs.

**Review by nflearl — CHANGES_REQUESTED** [review]
This PR does not follow the standard patterns for implementing cloneable.

The whole point of cloning is to not explicitly copy fields.  The only explicit need is to perform deep copies of referenced objects, which I did not find any examples in this review.

There is actually a more direct and surgical solution to the problem being solved.  The method getJsonString() effectively performs a deep copy of the event object into a Map which is then converted to a JSONObject.  That method is only called in 2 places one of which is the unit test.  If that method's signature is changed, then the User may be overridden after conversion instead of before.  This avoids the need to change LoggableEvent (and all the subclasses) entirely.

To convert from a JSONObject to a string, the getJsonString() method does nothing more than calling toString() which may be done anywhere.

Line 86 In the call method in KinesisEventPRocessor may be replaced as follows:

            json = getJsonString(event) + "\n";

could be replaced by

	JsonObject jsonObj = getJson(event);
	jsonObject.put("userKey", computeNewUserValue("oldUserValue"));
	json = jsonObj.toString() + "\n";

This code is just a sketch but should serve to illustrate a simpler and more direct approach, affecting just KinesisEventProcessor and perhaps its companion test class.

**Review by wnyffenegger — APPROVED** [review]

**Review by sb-benohe — APPROVED** [review]

**Review by nflearl — APPROVED** [review]

**Pull request merged**